dependency-specification.rst

```diff
@@ -0,0 +1,324 @@
+acceptable, so the language permits describing all these cases.
+
+The language defined is a compact line based format which is already in
+widespread use in pip requirements files, though we do not specify the command
+line option handling that those files permit. There is one caveat - the
+URL reference form, specified in PEP-440 [#pep440]_ is not actually
+implemented in pip, but since PEP-440 is accepted, we use that format rather
+than pip's current native format.
+
+Motivation
+==========
+
+Any specification in the Python packaging ecosystem that needs to consume
+lists of dependencies needs to build on an approved PEP for such, but
+PEP-426 [#pep426]_ is mostly aspirational - and there are already existing
+implementations of the dependency specification which we can instead adopt.
+The existing implementations are battle proven and user friendly, so adopting
+them is arguably much better than approving an aspirational, unconsumed, format.
+
+Specification
+=============
+
+Examples
+--------
+
+All features of the language shown with a name based lookup::
+
+    requests [security,tests] >= 2.8.1, == 2.8.* ; python_version < "2.7.10"
+
+A minimal URL based lookup::
+
+    pip @ https://github.com/pypa/pip/archive/1.3.1.zip#sha1=da9234ee9982d4bbb3c72346a6de940a148ea686
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. The grammar covers just the specification itself.
+The expectation is that the specification will be embedded into a larger
+system which offers framing such as comments, multiple line support via
+continuations, or other such features.
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">"
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    version-inner = version-cmp version *(',' version-cmp version)
+    versionspec   = ("(" version-inner ")") / version-inner
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
+    python-str    = "'" *(python-str-c / DQUOTE) "'"
+    python-str    =/ DQUOTE *(python-str-c / "'") DQUOTE
+    marker-var    = python-str / "python_version" / "python_full_version" /
```

Comment:
Your example ABNF would allow `platform_version in "Oracle"` but not `"Oracle" in platform_version`. Could fix that by adding some rules so `in` and `not in` require a literal on the left and variable on the right, while the other operators are the other way around, but dunno if it's worth it.